### PR TITLE
add support for single channel images

### PR DESCRIPTION
**Pull request by edgarriba**

### CODE REVIEW

File: torchvision/transforms.py [hunk 1]
```diff
@@ -43,7 +43,10 @@ class ToTensor(object):
     def __call__(self, pic):
         if isinstance(pic, np.ndarray):
             # handle numpy array
-            img = torch.from_numpy(pic.transpose((2, 0, 1)))
+            if len(pic.shape) >= 3:
```
Comment (fmassa) [inline torchvision/transforms.py]: You can replace `len(pic.shape)` by `pic.ndim`. Also, if the number of dimensions is > 3, the following line (with the `transpose`) will crash.
Maybe you could do something like
```python
if pic.ndim == 2:
    pic = pic[np.newaxis]
elif pic.ndim == 3:
    pic = pic.transpose((2, 0, 1))
else:
    raise ValueError "only 2D and 3D images accepted, got {}D image".format(pic.ndim)
img = torch.from_numpy(pic)
```

File: torchvision/transforms.py [hunk 2]
```diff
@@ -43,7 +43,10 @@ class ToTensor(object):
...
+                img = torch.from_numpy(pic.transpose((2, 0, 1)))
+            else:
+                img = torch.from_numpy(pic.reshape((1,) + pic.shape))
```
Comment (fmassa) [inline torchvision/transforms.py]: nit: might be simpler to do `pic[np.newaxis]` to append a new dimension into the array

File: torchvision/transforms.py [hunk 3]
```diff
@@ -89,6 +92,8 @@ def __call__(self, pic):
         if torch.is_tensor(pic):
             npimg = np.transpose(pic.numpy(), (1, 2, 0))
         assert isinstance(npimg, np.ndarray), 'pic should be Tensor or ndarray'
+        if len(npimg.shape) < 3:
```
Comment (fmassa) [inline torchvision/transforms.py]: nit: `npimg.ndim < 3` might be better?

File: torchvision/transforms.py [hunk 4]
```diff
@@ -89,6 +92,8 @@ def __call__(self, pic):
...
+            npimg = np.reshape(npimg, npimg.shape + (1,))
```
Comment (fmassa) [inline torchvision/transforms.py]: nit: `npimage[..., np.newaxis]`?

**Comment by alykhantejani** [comment]
@edgarriba are you able to make the changes requested by @fmassa?
Also, it would be good to add another case to `test_to_tensor` to test for single channel images.

**Comment by edgarriba** [comment]
@alykhantejani @fmassa sure, I'll allocate some time during this weekend

File: test/test_transforms.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import torch
 import torchvision.transforms as transforms
 import unittest
+from parameterized import parameterized
```
Comment (edgarriba) [inline test/test_transforms.py]: @alykhantejani @fmassa I've added this dependency which I think that is useful to handle different test cases

File: torchvision/transforms.py [hunk 6]
```diff
@@ -52,33 +52,38 @@ def __call__(self, pic):
...
+		pic.transpose((2, 0, 1))
+	    else:
+		raise ValueError('only 2D and 3D images accepted, got {}D image'.format(pic.ndim)
             # backward compatibility
-            return img.float().div(255)
+            return torch.from_numpy(pic).float().div(255.)
 
         if accimage is not None and isinstance(pic, accimage.Image):
-            nppic = np.zeros([pic.channels, pic.height, pic.width], dtype=np.float32)
+            nppic=np.zeros([pic.channels, pic.height, pic.width], dtype=np.float32)
```
Comment (alykhantejani) [inline torchvision/transforms.py]: PEP recommendations are to surround assignment operators with a space (except for in kwargs) - not sure why the linter didn't pick this up but this is how we've done it across the codebase, so I'd rather not change it in this file.

File: test/test_transforms.py (hunk 5 above)
Comment (alykhantejani) [reply test/test_transforms.py]: Not sure that we want to add another dependency. In the main `pytorch` repo these cases are handled by looping over the various parameterizations. It's not ideal, but does keep the dependency count down

File: test/test_transforms.py [hunk 7]
```diff
@@ -171,17 +172,35 @@ def test_lambda(self):
...
-    def test_to_tensor(self):
-        channels = 3
-        height, width = 4, 4
-        trans = transforms.ToTensor()
-        input_data = torch.ByteTensor(channels, height, width).random_(0, 255).float().div_(255)
+    @parameterized.expand([
+        ('3channel', 3, 4, 4),
+        ('1channel', 1, 4, 4),
+    ])
+    def test_pil_to_tensor(self, _, channels, height, width):
```
Comment (alykhantejani) [inline test/test_transforms.py]: nit: `test_pil_image_to_tensor`

**Comment by alykhantejani** [comment]
@edgarriba Thanks for this - in general it looks good just a few small things to fix. I don't think we want to add another dependency to the library if we can avoid it. Would you be able to make this change and the linting ones?

**Comment by utkuozbulak** [comment]
Checks for image/numpy channels are implemented with refactoring of transforms to functional and transforms in #311  

https://github.com/pytorch/vision/blob/50b2f910490a731c4cd50db5813b291860f02237/torchvision/transforms/functional.py#L43-L44

Although single channel images  (HxW) still produce following error: **ValueError: axes don't match array**  due to 
https://github.com/pytorch/vision/blob/50b2f910490a731c4cd50db5813b291860f02237/torchvision/transforms/functional.py#L46-L48

Wouldn't it be easier to just expand these arrays with a simple if, like:
```python
...
    if isinstance(pic, np.ndarray):  ## Existing
        # handle numpy array
        if pic.ndim == 2:  ## Addition
            # expand by one dimension as the last channel  ## Addition
        img = torch.from_numpy(pic.transpose((2, 0, 1)))  ## Existing
...
```

Either way, this PR might as well be closed as a clean up?

**Comment by fmassa** [comment]
I'm planning on modifying the underlying data structure that we use internally (instead of using a PIL image, have a custom class), so that those corner cases can be fixed.
I'll open an issue with the description of the proposal once I get the chance.

**Comment by utkuozbulak** [comment]
Sounds like a plan, I would like to work on that if you will need a pair of hands.

**Pull request closed without merging**